src/chat/tools/listTopics.ts

```diff
@@ -0,0 +1,159 @@
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No environment ID provided."),
+      ]);
+    }
+
+    const kafkaClusterId = params.kafkaClusterId;
+    if (!kafkaClusterId) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No Kafka cluster ID provided."),
+      ]);
+    }
+
+    const loader = ResourceLoader.getInstance(connectionId);
+
+    const kafkaClusters: KafkaCluster[] =
+      await loader.getKafkaClustersForEnvironmentId(environmentId);
+
+    if (!(Array.isArray(kafkaClusters) && kafkaClusters.length)) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No Kafka clusters found for the given environment ID."),
+      ]);
+    }
+    const kafkaCluster = kafkaClusters.find((cluster) => cluster.id === kafkaClusterId);
+    if (!kafkaCluster) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No Kafka cluster found for the given ID."),
+      ]);
+    }
+    const topics = await loader.getTopicsForCluster(kafkaCluster);
+    if (!(Array.isArray(topics) && topics.length)) {
+      logger.debug(`No topics found for cluster ${kafkaClusterId}`);
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No topics found for the given Kafka cluster."),
+      ]);
+    }
+
+    let sampleTopics: KafkaTopic[] = [...topics]; // Create a copy of the topics array
+    const topicNameSubstring = params.topicNameSubstring;
+
+    if (topicNameSubstring) {
+      sampleTopics = sampleTopics.filter((topic) =>
+        topic.name.toLowerCase().includes(topicNameSubstring.toLowerCase()),
+      );
+      logger.debug(
+        `Filtered topics by substring "${topicNameSubstring}": ${sampleTopics.length} matches`,
+      );
+    }
+
+    // if sample topics is more than 30, slice it to 30
+    if (sampleTopics.length > 30) {
+      sampleTopics = sampleTopics.slice(0, 30);
+      logger.debug(`Limited topics to 30 samples`);
+    }
+
+    if (token.isCancellationRequested) {
+      logger.debug("Tool invocation cancelled");
+      return new LanguageModelToolResult([]);
+    }
+
+    logger.debug(`Summarizing ${sampleTopics.length} topics`);
+    return new LanguageModelToolResult(summarizeTopics(sampleTopics));
+  }
+  async processInvocation(
+    request: ChatRequest,
+    stream: ChatResponseStream,
+    toolCall: LanguageModelToolCallPart,
+    token: CancellationToken,
+  ): Promise<TextOnlyToolResultPart> {
+    const parameters = toolCall.input as IListTopicsParameters;
+    const progressMessage = [
+      "Retrieving available topics for:",
+      `- Cluster ID: ${parameters.kafkaClusterId}`,
+      `- Connection ID: ${parameters.connectionId}`,
+      `- Environment ID: ${parameters.environmentId}`,
+    ];
+
+    if (parameters.topicNameSubstring) {
+      progressMessage.push(`- Filter: "${parameters.topicNameSubstring}"`);
+    }
+
+    stream.progress(progressMessage.join("\n"));
```

Comment:
While this is interesting to see, I think we should reduce the amount of information shown here a bit -- even with just the local connection and a single direct connection, this is taking up 3+ progress rows' worth of space. 😅 
<img width="849" alt="image" src="https://github.com/user-attachments/assets/3ac2474a-d9ba-4fc8-9b90-4d865881a484" />
(Plus we don't know if VS Code will continue to support markdown-style text here)